### PR TITLE
[proof] Removal of global proof state.

### PR DESCRIPTION
**Pull request by ejgallego**
Currently, proofs are stored in a global variable in `Proof_global`,
which is then kept in sync with the `Stm` notion of proof for each
particular state.

In this commit, we make some partial progress towards removal of such
global state in favor of a functional handling of the proof state. For
now, we deprecate functions accessing the global proof state to get an
idea of the remaining work.

Completing this work seems doable for 8.8, however the question on
compatibility should be addressed. There are 3 possible choices:

- Deprecate all the interfaces, but remove the global proof state in
  8.9 . This implies having Coq use some of the bad interfaces itself.

- Hard removal of the impure interfaces.

- Removal of the global proof state, keep the deprecated interfaces
  and install a private compatibility hook for the STM to update the
  proof state while the transition happens.

I think the best option is 2, even if a bit painful. 1 has a big risk
of delaying work and so the work may never happen. 3 is a bit of a
hack and IMO the advantages are not so clear.

EDIT: this depends on #7196.

### CODE REVIEW

File: plugins/ltac/coretactics.ml4
Comment (herbelin) [inline plugins/ltac/coretactics.ml4]: I prefer #7196 :smile:

**Review by herbelin — COMMENTED** [review]
This is a pretty difficult PR to review into details, but "hard removal of the impure interfaces" seems also to me the most sustainable direction to follow. I'm ready to trust by default the current work done in this direction.

**Review by forestjulien — APPROVED** [review]

**Review by Zimmi48 — COMMENTED** [review]
There are a few user error messages that were associated with the exceptions that are removed. I'd like to understand what they become with this PR: what are the error messages now printed to the user with the input that was previously leading to these exceptions.

**Comment by ejgallego** [comment]
> There are a few user error messages that were associated with the exceptions that are removed. I'd like to understand what they become with this PR: what are the error messages now printed to the user with the input that was previously leading to these exceptions.

That's a very good point, as far as I can see these messages were all mostly duplicated, the plan is indeed to centralize the commands that require a proof to operate an unify this error in a single place.

Maybe we could indeed refine the vernacular for this, but there are 3 kinds: commands that don't or optionally access the proof environment, commands that access it read-only, and commands that access and modify it. We will see.

**Comment by ppedrot** [comment]
What's the ETA of this PR?

**Comment by ejgallego** [comment]
I plan to work on it this upcoming CIW. Hopefully I get a merge candidate, if not, it should be very close.

The only remaining thing to is to clean up `vernacentries`, as there is a lot of mess with regards to commands that use the proof state [optionally/mandatory] [read-only/read-write].

A person familiar with that part of the code is @vbgl , maybe he's interested in giving a hand?

**Review by ppedrot — APPROVED** [review]
FTR I support the idea and what is implemented currently looks OK.

**Comment by ppedrot** [comment]
@ejgallego Did you make any progress?

**Comment by ejgallego** [comment]
> @ejgallego Did you make any progress?

Unfortunately not and this PR is prioritary, next target to work on this end of month.

**Comment by ejgallego** [comment]
I am semi-orphaning this PR, as I have realized that the original motivation to have this cleanup in was naïve in restrospective (i.e: it will take decades to get all the proper changes in)

Note however that I think this cleanup would be indeed nice to complete, so I may try to give it a go in the next weeks, but feel free to take over. There is not a lot of work remaining to do.

**Comment by ppedrot** [comment]
I was OK with the current state of affairs. Any small progress is better than none, and on changes that are supposed to take decades it is probably the way to go.

**Comment by ejgallego** [comment]
> changes that are supposed to take decades it is probably the way to go.

Changes will take decades as opposed to weeks not because they are so complex, but because 99% of the development time is spent waiting for reviews/merges/bike-shedding.

**Comment by ejgallego** [comment]
Note that I have a new version of this but it depends on quite a lot of deprecation PRs so it will have to wait until we clear those out as to avoid spamming.

**Comment by ejgallego** [comment]
> I feel that we are many to want that it does not take decades. Let's try to make the functionalization of the state a priority.

Thanks @herbelin; this is currently blocked on one of the hardest parts of the whole project: having the STM thread the state. Once we manage to do that we could indeed progress, but this is not going to be easy.

**Comment by herbelin** [comment]
Looking again at this PR, and in relation to the questions discussed in #8590, #8598, #8600, another question that comes to my mind is the following: you are threading the `pstate`, but the proof state is only one component of the collection of all individual states. For instance, at some time, `declare_fixpoint` will also return the implicit arguments, notations, short names, that it shall have declared. So, I guess you intend to have `declare_fixpoint` eventually returning a more complex data than just `Proof_global.t option`. So, why not go directly to a situation where functions like `declare_fixpoint` have type of the form:
```ocaml
val declare_fixpoint : gallina_state -> locality -> polymorphic ->
   recursive_preentry * UState.universe_decl * UState.t *
   (Constr.rel_context * Impargs.manual_implicits * int option) list ->
   Proof_global.lemma_possible_guards -> decl_notation list ->
   gallina_state
```
with the type `gallina_state` equipped with update/lookup methods relatively to `Proof_global.t option`?

Then, we could stepwise add more and more functionally-threaded individual states to `gallina_state`.

**Comment by ejgallego** [comment]
> you are threading the `pstate`, but the proof state is only one component of the collection of all individual states.

That's a very good question. The story here is that the "proof state" is actually managed outside the regular mechanisms in `State` and `Summary`, due to parsing / "proof modes" need. So indeed, what you say could indeed be done, but doing this bit seems a bit more prioritary as the current handling in the codebase seems to have several problems.

**Comment by ejgallego** [comment]
> I feel that we are many to want that it does not take decades.

Anyways my original comment was referring to the problem of needing a large amount of changes and having a huge delay for each of those. One example is #8501, so indeed 12 days for a trivial change to get in, and no perspective of the situation getting better. So at this rate we cannot be optimistic in that we will complete a project of this magnitude in any reasonable amount of time.

For reference, the changes I had to do in order to enable a compartmentalized build + Dune should have taken at most 1 week of development / review + 1 week for merging, instead they took several months. So yeah, it is not feasible to perform such kind of projects in a reasonable timeframe.

**Comment by Zimmi48** [comment]
I agree with you that some PRs are taking way too long to get merged (I haven't looked into stats yet, but it feels clear that it depends mainly on who the assignee / code owners are). However you should stop using Dune as an example because you were responsible of a large part of the months it took to finally get it in (from the moment I said I would get your Dune PRs merged, it still took you several months to get to a state where you were satisfied). You have plenty of better examples.

**Comment by ejgallego** [comment]
> However you should stop using Dune as an example because you were responsible of a large part of the months it took to finally get it in

I was thinking about the dependencies of the Dune PR itself [interfaces, binaries reorg] indeed, those were very painful to get in. Anyways sorry for the offtopic.

**Comment by herbelin** [comment]
So, if I understand correctly, we are only waiting on #8493, and this PR can be merged, so that next steps can then be done.

In passing, about the next steps: If we consider that each proof should come with its own "global state", the proof state (of type `Proof_global.t option`) should become the "master state" of the system.

**Comment by herbelin** [comment]
Also waiting a rebase and a CI confirmation.

**Comment by ejgallego** [comment]
> So, if I understand correctly, we are only waiting on #8493, and this PR can be merged, so that next steps can then be done.

@herbelin, indeed we are waiting on that PR, but also a few parts of the code remain to be adapted, mainly LTAC is the last thing. I don't know if I will find some strange problems.

> In passing, about the next steps: If we consider that each proof should come with its own "global state", the proof state (of type `Proof_global.t option`) should become the "master state" of the system.

This is actually the other way around, the "master" state owns the proof state, i.e: There can be many parts of the document where there is for example no open proof.

I suggest we replace the terminology "master" state with "document state", as master state feels confusing where in fact, given a document, each part of it features a different state.

**Comment by herbelin** [comment]
> > In passing, about the next steps: If we consider that each proof should come with its own "global state", the proof state (of type `Proof_global.t option`) should become the "master state" of the system.
> 
> This is actually the other way around, the "master" state owns the proof state, i.e: There can be many parts of the document where there is for example no open proof.

My wording was confusing. The "master" indeed owns the proof states but the proof states themselves own a `States.state`. This is this reversal of point of view which I wanted to highlight (but probably it was already clear for you and it just took me time to realize it).

> I suggest we replace the terminology "master" state with "document state", as master state feels confusing where in fact, given a document, each part of it features a different state.

Why not, sounds interesting.

**Comment by herbelin** [comment]
> Maybe I am not understanding what you mean, what happens then for the parts of the document that have no open proof. These still need a `States.state`.

Yes, this is what was confusing. There is a global `State.state` and a branching of extensions of it, one for each goal.

**Comment by ejgallego** [comment]
> There is a global `State.state` and a branching of extensions of it, one for each goal.

That doesn't seem to be the case, for example setting a printing option in one proof will alter the printing state of the rest of the proofs IIUC. See the code in `Proof_global`.

So indeed my point is that already we need to clean up things here and with `env` before starting to worry about the summary / libobject itself.

**Comment by herbelin** [comment]
> That doesn't seem to be the case, for example setting a printing option in one proof will alter the printing state of the rest of the proofs IIUC. See the code in `Proof_global`.

OK.

> So indeed my point is that already we need to clean up things here and with env 

Yes.

> before starting to worry about the summary / libobject itself.

But how to maintain a correct state for a goal without worring also about the summaries? E.g. in the case of `abstract`, we want that the subproof remains local to the state of the goal and does not interfere with the state of other goals, right? And we want that the state of the goal is backtracted if I use `abstract` in a `try`, right? How to have this locality if `cache_object` does an action to the single mutable global state. We then do want that `cache_object` works functionally, no? (Or do you indeed that we encapsulate `cache_object` between a `freeze` of the main global state to keep a copy, an `unfreeze` of the goal state, the call to `cache_object` - thus working on the goal state -, a `freeze` to update the goal state, an `unfreeze` to restore the main global state?)

Back to `Set option`, what locality/globality of the semantic would you like to see?

**Comment by herbelin** [comment]
> > But how to maintain a correct state for a goal without worrying also about the summaries?
> 
> This is a good question, what does the current code do today?

The current code has bugs, `abstract` modifies the global state but also adds a "side-effect" to be reported at the end. Backtracing on an `abstract` does not backtrack on the modification of the global state that `abstract` made. As of today, we need to have `abstract` modifies the global state because otherwise, e.g. simply `Show Proof`, works only with the global state, so data about the subproof, its entry in the nametab, its implicit arguments, its argument scopes, etc. have to be in the only state which `Show Proof` looks for which is the global state. Etc.

> What part of `States.state` does `abstract` modify?

Basically all parts that a `Definition` modifies: nametab, global env, implicit arguments, argument scopes, ...

> > Back to `Set option`, what locality/globality of the semantic would you like to see?
>
> What do you exactly mean? I think I am fine with the current one. Do you see a problem?

I meant that, in the case of printing options, there is the eternal dilemma of options being controlled by the GUI or by the document.

Also, it is not obvious to me that options have to survive the locality of sections and/or proofs.

**Comment by ejgallego** [comment]
> Basically all parts that a `Definition` modifies: nametab, global env, implicit arguments, argument scopes, ..

Should `abstract` modify that part?

> I meant that, in the case of printing options, there is the eternal dilemma of options being controlled by the GUI or by the document.

I see, I think both models are compatible; usually UI will have a better time sending the options on each print call [as to avoid races and weird stuff], but we can keep a default status too without too much trouble as is done today.

> Also, it is not obvious to me that options have to survive the locality of sections and/or proofs.

That seems a matter fairly orthogonal to the functional handling of state bug more like a matter for #8096

**Comment by herbelin** [comment]
> > Basically all parts that a `Definition` modifies: nametab, global env, implicit arguments, argument scopes, ..
> 
> Should `abstract` modify that part?

Even if we don't want a local subproof to obey the general rules for implicit arguments, argument scopes, etc, we still need to have it somewhere in some environment so that all functions which operate at some time on the proof knows what this subproof is. Do you have other ways to solve the `abstract` bugs? Are you thinking e.g. at simulating the presence of subproofs as a section context variable?

**Comment by ejgallego** [comment]
This PR is in dormant state until we can merge #9129 ; then we should proceed to merge the two extra commits this PR has.

**Comment by ejgallego** [comment]
> This PR is in dormant state until we can merge #9129 ; then we should proceed to merge the two extra commits this PR has.

Actually much has happened since this and this PR is obsolete.

**Pull request closed without merging**
